### ISSUE TITLE
Rule Proposal: `no-unused-labels`

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#5052]
This rule is similar to [no-unused-vars](http://eslint.org/docs/rules/no-unused-vars).
This rule warns unused labels.

The following patterns are considered problems:

```js
/*eslint no-unused-labels: 2*/

A: {                               /*error 'A:' is defined but never used.*/
    foo();
}

B: for (let i = 0; i < 10; ++i) {  /*error 'B:' is defined but never used.*/
    foo();
}

// I hope, in cases that we can use unlabeled break/continue.
C:                                 /*error 'C:' is defined but not needed.*/
for (let i = 0; i < 10; ++i) {
    if (foo()) {
        break C;
    }
    bar();
}
```

The following patterns are not considered problems:

```js
/*eslint no-unused-labels: 2*/

A: {
    if (foo()) {
        break A;
    }
    bar();
}

B:
for (let i = 0; i < 10; ++i) {
    for (let j = 0; j < 10; ++j) {
        if (foo()) {
            break B;
        }
    }
    bar();
}
```

**Comment by mysticatea** [comment]
This rule may be covering the domain of [no-empty-label](http://eslint.org/docs/rules/no-empty-label). Because the "empty" labels cannot be used.

**Comment by michaelficarra** [comment]
With the C case, the label is redundant, but cannot be naively removed without also changing the break statement. @nzakas might have a problem with that.

**Comment by alberto** [comment]
Maybe there should be a specific `no-unneded-label` rule to error on `break C` and leave it out from this one.

**Comment by nzakas** [comment]
Yeah, might be best to break up cases A and B from C, just to be clear. The error on C seems like it would be on the `break` rather than the label.

**Comment by mysticatea** [comment]
OK, I will split out the C case from this proposal.


The following patterns are considered problems:

```js
/*eslint no-unused-labels: 2*/

A: {                               /*error 'A:' is defined but never used.*/
    foo();
}

B: for (let i = 0; i < 10; ++i) {  /*error 'B:' is defined but never used.*/
    foo();
}
```

The following patterns are not considered problems:

```js
/*eslint no-unused-labels: 2*/

A: {
    if (foo()) {
        break A;
    }
    bar();
}

B:
for (let i = 0; i < 10; ++i) {
    if (foo()) {
        break B;
    }
    bar();
}
```

**Comment by nzakas** [comment]
Looks good. :+1:

**Comment by mysticatea** [comment]
I'm working on this.
